### PR TITLE
operator: Add network tenant for netobserv

### PR DESCRIPTION
**Pull request by jpinsonneau**
**What this PR does / why we need it**:
This PR adds a `network` tenant in Openshift Gateway component for [netobserv](https://github.com/netobserv) project. This will allow [network-observability-operator](https://github.com/netobserv/network-observability-operator) to configure its components to read and write `loki-operator` logs.

**Which issue(s) this PR fixes**:
/

**Special notes for your reviewer**:
@periklis I feel recommending users to use another lokistack instance is good enough for now but feel free if you feel we should split `ModeType` or add additional guard.

**Checklist**
- [X] Documentation added
- [X] Tests updated
- [ ] Is this an important fix or new feature? Add an entry in the `CHANGELOG.md`.
- [ ] Changes that require user attention or interaction to upgrade are documented in `docs/sources/upgrading/_index.md`

### CODE REVIEW

**Comment by jpinsonneau** [comment]
As discussed with @periklis I have updated this PR to add a new `openshift-network` mode holding the `network` tenant.
This ensure users split their instances according to their needs.

File: operator/docs/user-guides/forwarding_logs_to_gateway.md [hunk 1]
```diff
@@ -134,6 +134,59 @@ _Note: While this document will only give instructions for two methods of log fo
...
+      apiGroup: rbac.authorization.k8s.io
+      kind: ClusterRole
+      name: lokistack-network-tenant-logs
+    subjects:
+    - kind: ServiceAccount
+      name: flowlogs-pipeline
+      namespace: network-observability
+    - kind: ServiceAccount
+      name: network-observability-plugin
+      namespace: network-observability
```
Comment (periklis) [inline operator/docs/user-guides/forwarding_logs_to_gateway.md:176]: Do you really use for a namespace with the `openshift-` prefix in the name, i.e. `openshift-network-observability`?

File: operator/docs/user-guides/forwarding_logs_to_gateway.md [hunk 2]
```diff
@@ -134,6 +134,59 @@ _Note: While this document will only give instructions for two methods of log fo
...
+* Deploy the [Network Observability Operator](https://github.com/netobserv/network-observability-operator) following the [Getting started documentation](https://github.com/netobserv/network-observability-operator/blob/main/README.md#getting-started) either from Operator Hub or commands available in the repository.
+
+* Apply the following configuration in `FlowCollector` for `network` tenant of `lokistack-network`:
+```yaml
+  loki:
+    tenantID: network
+    sendAuthToken: true
+    url: 'https://lokistack-network-gateway-http.openshift-logging.svc.cluster.local:8080/api/logs/v1/network/'
+```
+Check [config samples](https://github.com/netobserv/network-observability-operator/tree/main/config/samples) for all options.
```
Comment (periklis) [inline operator/docs/user-guides/forwarding_logs_to_gateway.md:188]: Per our live discussion I had the notion that you are not going to re-use the LokiStack instance on `openshift-logging` but rather install your LokiStack instance` on `network-observability` ns. Right?

If I am right I suggest to add steps to install the LokiStack manually on `network-observability` ns here.

File: operator/internal/handlers/internal/gateway/tenant_configmap_test.go [hunk 3]
```diff
@@ -80,6 +85,11 @@ func TestGetTenantConfigMapData_ConfigMapExist(t *testing.T) {
 				CookieSecret: "test789",
 			},
 		},
+		"network": {
+			OpenShift: &manifests.TenantOpenShiftSpec{
+				CookieSecret: "test101112",
+			},
+		},
```
Comment (periklis) [inline operator/internal/handlers/internal/gateway/tenant_configmap_test.go:92]: I don't believe this test case adds more value as it only adds an additional tenant to the stub and expects to find it after calling `GetTenantConfigMapData`. There is nothing `network`-tenant-specific added to the `GetTenantConfigMapData` itself.

File: operator/internal/manifests/internal/gateway/gateway-tenants.yaml [hunk 4]
```diff
@@ -79,5 +79,19 @@ tenants:
...
+  id: {{ $spec.TenantID }}
+  openshift:
+    serviceAccount: {{ $spec.ServiceAccount }}
+    redirectURL: {{ $spec.RedirectURL }}
+    cookieSecret: {{ $spec.CookieSecret }}
+  opa:
+    url: {{ $l.OpenShiftOptions.Authorization.OPAUrl }}
+    withAccessToken: true
+{{- end -}}
+{{- end -}}
```
Comment (periklis) [inline operator/internal/manifests/internal/gateway/gateway-tenants.yaml:95]: I suggest to collapse this with the block for `openshift-logging` and make the guarding else if to accept `openshift-logging` or `openshift-network` as tenants mode.

File: operator/internal/manifests/internal/gateway/build_test.go [hunk 5]
```diff
@@ -195,6 +195,15 @@ tenants:
...
     withAccessToken: true
+- name: network
+  id: 3e922593-e352-47df-8c5c-c39dbdd5b83c
+  openshift:
+    serviceAccount: lokistack-gateway
+    redirectURL: https://localhost:8443/openshift/network/callback
+    cookieSecret: whynot
+  opa:
+    url: http://127.0.0.1:8080/v1/data/lokistack/allow
+    withAccessToken: true
```
Comment (periklis) [inline operator/internal/manifests/internal/gateway/build_test.go:206]: I strongly believe this should be a separate test case where you separately configure the `Options.Stack.Tenants.Mode: lokiv1.OpenShiftNetwork` and you get a config rendering with only a single tenant namely `network`.

**Review by periklis — CHANGES_REQUESTED** [review]
Looks in general good to me. I added a couple of suggestions only to make tests for tenant `openshift-network` separate. Besides I would like to clarify that you will start by installing a custom LokiStack in the `network-observability` namespace. If yes, please assert that you get ClusterRole/ClusterRoleBindings for you instance properly reconciled. I am pretty sure this is the case but nevermind double checking.

File: operator/internal/handlers/internal/gateway/tenant_configmap_test.go (hunk 3 above)
Comment (jpinsonneau) [reply operator/internal/handlers/internal/gateway/tenant_configmap_test.go:92]: Sure, I can remove that part

File: operator/internal/manifests/internal/gateway/gateway-tenants.yaml (hunk 4 above)
Comment (jpinsonneau) [reply operator/internal/manifests/internal/gateway/gateway-tenants.yaml:95]: Agreed. I was not really comfortable on that part. Could you please help me on this ?

Comment (periklis) [reply operator/internal/manifests/internal/gateway/gateway-tenants.yaml:95]: Sure you change L68 to something like that:
```go
{{- else if (or (eq $l.Stack.Tenants.Mode "openshift-logging") (eq $l.Stack.Tenants.Mode "openshift-network")) -}}
```

**Comment by jpinsonneau** [comment]
@periklis thank you for your support !

I did the requested changes. Feel free if you have any concerns.

File: operator/docs/user-guides/forwarding_logs_to_gateway.md (hunk 2 above)
Comment (jpinsonneau) [reply operator/docs/user-guides/forwarding_logs_to_gateway.md]: I have added the manual steps, however, the target for GA is an automatic installation and configuration of Network Observability dependent operators (loki, kafka, prometheus, etc)

**Review by periklis — COMMENTED** [review]
Looks great! Let me pass another full review on the next work day and we are set to go.

**Review by periklis — APPROVED** [review]

File: operator/README.md [hunk 6]
```diff
@@ -25,7 +25,7 @@ Do not use this in any production environment.**
 
 #### Sending Logs Directly to the Distributor Component
 
-* The [forwarding logs to LokiStack without LokiStack Gateway](https://github.com/grafana/loki/tree/master/operator/docs/forwarding_logs_without_gateway.md) is used to send application, infrastructure, and audit logs to the Loki Distributor as different tenants using Fluentd or Vector.
+* The [forwarding logs to LokiStack without LokiStack Gateway](https://github.com/grafana/loki/tree/master/operator/docs/forwarding_logs_without_gateway.md) is used to send application, infrastructure, audit and network logs to the Loki Distributor as different tenants using Fluentd or Vector.
 * The guide has a step-by-step guide to connect with OpenShift Logging for forwarding logs to LokiStack.
```
Comment (periklis) [inline operator/README.md:29]:
```suggestion
* The guide has a step-by-step guide to connect with OpenShift Logging or OpenShift Network for forwarding logs to LokiStack.
```

**Review by periklis — APPROVED** [review]

**Pull request merged**